File: media-server/load_test.py
#!/usr/bin/env python3
"""
Media Server Load Testing Script

Simulates heavy concurrent load to validate architectural improvements:
- Template serialization bottleneck elimination
- Containerization readiness
- API endpoint performance
- Random mode navigation under load

Usage:
    python3 load_test.py --url http://localhost:8080 --workers 5 --duration 60
    python3 load_test.py --url http://localhost:8080 --scenario stress --workers 10
    python3 load_test.py --url http://localhost:8080 --scenario api-only --requests 1000
"""

import argparse
import concurrent.futures
import json
import random
import sys
import time
from collections import defaultdict
from dataclasses import dataclass, field
from datetime import datetime, timedelta
from typing import Dict, List, Optional
from urllib.parse import quote

import requests


@dataclass
class RequestStats:
    """Statistics for a single request"""
    endpoint: str
    duration_ms: float
    status_code: int
    error: Optional[str] = None
    timestamp: float = field(default_factory=time.time)


@dataclass
class TestResults:
    """Aggregate test results"""
    total_requests: int = 0
    successful_requests: int = 0
    failed_requests: int = 0
    broken_pipes: int = 0
    timeouts: int = 0
    total_duration_ms: float = 0.0
    min_duration_ms: float = float('inf')
    max_duration_ms: float = 0.0
    errors_by_type: Dict[str, int] = field(default_factory=lambda: defaultdict(int))
    requests_by_endpoint: Dict[str, int] = field(default_factory=lambda: defaultdict(int))
    stats: List[RequestStats] = field(default_factory=list)

    def add_request(self, stat: RequestStats):
        """Add a request statistic"""
        self.stats.append(stat)
        self.total_requests += 1
        self.requests_by_endpoint[stat.endpoint] += 1

        if stat.error:
            self.failed_requests += 1
            self.errors_by_type[stat.error] += 1
            if 'broken pipe' in stat.error.lower():
                self.broken_pipes += 1
            if 'timeout' in stat.error.lower():
                self.timeouts += 1
        else:
            self.successful_requests += 1
            self.total_duration_ms += stat.duration_ms
            self.min_duration_ms = min(self.min_duration_ms, stat.duration_ms)
            self.max_duration_ms = max(self.max_duration_ms, stat.duration_ms)

    def avg_duration_ms(self) -> float:
        """Calculate average duration"""
        if self.successful_requests == 0:
            return 0.0
        return self.total_duration_ms / self.successful_requests

    def success_rate(self) -> float:
        """Calculate success rate percentage"""
        if self.total_requests == 0:
            return 0.0
        return (self.successful_requests / self.total_requests) * 100

    def requests_per_second(self, elapsed_seconds: float) -> float:
        """Calculate throughput"""
        if elapsed_seconds == 0:
            return 0.0
        return self.total_requests / elapsed_seconds


class LoadTester:
    """Load testing orchestrator"""

    def __init__(self, base_url: str, timeout: int = 30):
        self.base_url = base_url.rstrip('/')
        self.timeout = timeout
        self.session = requests.Session()
        self.categories: List[str] = []
        self.files_by_category: Dict[str, List[str]] = {}

    def discover_categories(self) -> List[str]:
        """Fetch available categories from server"""
        try:
            response = self.session.get(f"{self.base_url}/", timeout=self.timeout)
            # Parse homepage HTML to extract category names
            # For now, use common test categories
            self.categories = ['All', '5-★★★★★', '📁 Classico']
            print(f"✓ Using test categories: {self.categories}")
            return self.categories
        except Exception as e:
            print(f"⚠ Failed to discover categories: {e}")
            self.categories = ['All']
            return self.categories

    def fetch_file_list(self, category: str) -> List[str]:
        """Fetch file list for a category via API"""
        try:
            encoded_category = quote(category)
            url = f"{self.base_url}/api/filelist?category={encoded_category}"
            response = self.session.get(url, timeout=self.timeout)

            if response.status_code == 200:
                files = response.json()
                self.files_by_category[category] = files
                print(f"✓ Fetched {len(files)} files for category '{category}'")
                return files
            else:
                print(f"⚠ Failed to fetch files for '{category}': {response.status_code}")
                return []
        except Exception as e:
            print(f"⚠ Error fetching file list for '{category}': {e}")
            return []

    def make_request(self, endpoint: str, params: Dict = None) -> RequestStats:
        """Make a single HTTP request and record statistics"""
        start = time.time()
        url = f"{self.base_url}{endpoint}"

        try:
            response = self.session.get(url, params=params, timeout=self.timeout)
            duration_ms = (time.time() - start) * 1000

            return RequestStats(
                endpoint=endpoint,
                duration_ms=duration_ms,
                status_code=response.status_code,
                error=None if response.status_code < 400 else f"HTTP {response.status_code}"
            )
        except requests.exceptions.Timeout:
            duration_ms = (time.time() - start) * 1000
            return RequestStats(
                endpoint=endpoint,
                duration_ms=duration_ms,
                status_code=0,
                error="Timeout"
            )
        except requests.exceptions.ConnectionError as e:
            duration_ms = (time.time() - start) * 1000
            error_msg = "Broken pipe" if "Broken pipe" in str(e) else "Connection error"
            return RequestStats(
                endpoint=endpoint,
                duration_ms=duration_ms,
                status_code=0,
                error=error_msg
            )
        except Exception as e:
            duration_ms = (time.time() - start) * 1000
            return RequestStats(
                endpoint=endpoint,
                duration_ms=duration_ms,
                status_code=0,
                error=str(e)
            )

    def viewer_request(self, category: str, random_mode: bool = False) -> RequestStats:
        """Make a viewer page request"""
        files = self.files_by_category.get(category, [])
        if not files:
            # Fallback - just request the category viewer without specific file
            return self.make_request(f"/view/{quote(category)}")

        # Pick a random file from the category
        file_path = random.choice(files)
        params = {'file': file_path}
        if random_mode:
            params['random'] = 'true'

        return self.make_request(f"/view/{quote(category)}", params)

    def gallery_request(self, category: str, page: int = 1) -> RequestStats:
        """Make a gallery page request"""
        return self.make_request(f"/tag/{quote(category)}", {'page': page})

    def api_filelist_request(self, category: str) -> RequestStats:
        """Make an API file list request"""
        return self.make_request(f"/api/filelist", {'category': category})

    def homepage_request(self) -> RequestStats:
        """Make a homepage request"""
        return self.make_request("/")

    def mixed_workload(self) -> RequestStats:
        """Execute a random mixed workload"""
        if not self.categories:
            return self.homepage_request()

        category = random.choice(self.categories)
        workload_type = random.choices(
            ['viewer', 'gallery', 'api', 'homepage'],
            weights=[40, 30, 20, 10]  # Weighted distribution
        )[0]

        if workload_type == 'viewer':
            return self.viewer_request(category, random_mode=random.choice([True, False]))
        elif workload_type == 'gallery':
            return self.gallery_request(category, page=random.randint(1, 5))
        elif workload_type == 'api':
            return self.api_filelist_request(category)
        else:
            return self.homepage_request()


def worker_thread(tester: LoadTester, scenario: str, duration_seconds: int,
                  requests_count: Optional[int] = None) -> List[RequestStats]:
    """Worker thread that executes requests"""
    results = []
    start_time = time.time()
    request_count = 0

    while True:
        # Check termination conditions
        elapsed = time.time() - start_time
        if duration_seconds > 0 and elapsed >= duration_seconds:
            break
        if requests_count and request_count >= requests_count:
            break

        # Execute request based on scenario
        if scenario == 'mixed':
            stat = tester.mixed_workload()
        elif scenario == 'viewer-random':
            category = random.choice(tester.categories) if tester.categories else 'All'
            stat = tester.viewer_request(category, random_mode=True)
        elif scenario == 'gallery':
            category = random.choice(tester.categories) if tester.categories else 'All'
            stat = tester.gallery_request(category)
        elif scenario == 'api-only':
            category = random.choice(tester.categories) if tester.categories else 'All'
            stat = tester.api_filelist_request(category)
        elif scenario == 'stress':
            # Aggressive stress test - no delays
            stat = tester.mixed_workload()
        else:
            stat = tester.mixed_workload()

        results.append(stat)
        request_count += 1

        # Add small delay for non-stress scenarios
        if scenario != 'stress' and duration_seconds > 0:
            time.sleep(0.1)

    return results


def print_results(results: TestResults, elapsed_seconds: float):
    """Print formatted test results"""
    print("\n" + "="*80)
    print("LOAD TEST RESULTS")
    print("="*80)

    print(f"\n📊 Overall Statistics:")
    print(f"  Total Requests:      {results.total_requests:,}")
    print(f"  Successful:          {results.successful_requests:,} ({results.success_rate():.2f}%)")
    print(f"  Failed:              {results.failed_requests:,}")
    print(f"  Broken Pipes:        {results.broken_pipes:,} {'⚠️  BOTTLENECK!' if results.broken_pipes > 0 else '✅'}")
    print(f"  Timeouts:            {results.timeouts:,}")
    print(f"  Test Duration:       {elapsed_seconds:.2f}s")
    print(f"  Throughput:          {results.requests_per_second(elapsed_seconds):.2f} req/s")

    if results.successful_requests > 0:
        print(f"\n⏱️  Response Times:")
        print(f"  Average:             {results.avg_duration_ms():.2f} ms")
        print(f"  Min:                 {results.min_duration_ms:.2f} ms")
        print(f"  Max:                 {results.max_duration_ms:.2f} ms")

        # Calculate percentiles
        durations = sorted([s.duration_ms for s in results.stats if not s.error])
        if durations:
            p50 = durations[int(len(durations) * 0.50)]
            p95 = durations[int(len(durations) * 0.95)]
            p99 = durations[int(len(durations) * 0.99)]
            print(f"  p50:                 {p50:.2f} ms")
            print(f"  p95:                 {p95:.2f} ms")
            print(f"  p99:                 {p99:.2f} ms")

    if results.requests_by_endpoint:
        print(f"\n🎯 Requests by Endpoint:")
        for endpoint, count in sorted(results.requests_by_endpoint.items(), key=lambda x: x[1], reverse=True):
            print(f"  {endpoint:30} {count:,}")

    if results.errors_by_type:
        print(f"\n❌ Errors by Type:")
        for error_type, count in sorted(results.errors_by_type.items(), key=lambda x: x[1], reverse=True):
            print(f"  {error_type:30} {count:,}")

    print("\n" + "="*80)

    # Verdict
    if results.broken_pipes > 0:
        print("⚠️  VERDICT: BROKEN PIPES DETECTED - Template serialization bottleneck still present!")
    elif results.success_rate() >= 99.0 and results.avg_duration_ms() < 500:
        print("✅ VERDICT: EXCELLENT - System handles load gracefully!")
    elif results.success_rate() >= 95.0:
        print("✓ VERDICT: GOOD - System performs adequately under load")
    else:
        print("⚠️  VERDICT: POOR - High error rate indicates issues")
    print("="*80 + "\n")


def main():
    parser = argparse.ArgumentParser(
        description="Load test the media server",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Scenarios:
  mixed          Mixed workload (default) - homepage, gallery, viewer, API
  viewer-random  Focus on viewer with random mode (stress test serialization)
  gallery        Gallery page requests with pagination
  api-only       API endpoint testing (/api/filelist)
  stress         Aggressive stress test with no delays

Examples:
  # Standard mixed load test with 5 workers for 60 seconds
  python3 load_test.py --url http://localhost:8080 --workers 5 --duration 60

  # Stress test viewer with random mode (tests serialization bottleneck)
  python3 load_test.py --url http://localhost:8080 --scenario viewer-random --workers 10 --duration 30

  # API endpoint performance test
  python3 load_test.py --url http://localhost:8080 --scenario api-only --requests 1000

  # Maximum stress test
  python3 load_test.py --url http://localhost:8080 --scenario stress --workers 20 --duration 120
        """
    )

    parser.add_argument('--url', default='http://localhost:8080',
                        help='Base URL of media server (default: http://localhost:8080)')
    parser.add_argument('--workers', type=int, default=5,
                        help='Number of concurrent workers (default: 5)')
    parser.add_argument('--duration', type=int, default=60,
                        help='Test duration in seconds (default: 60, 0 for request count only)')
    parser.add_argument('--requests', type=int, default=None,
                        help='Number of requests per worker (overrides duration)')
    parser.add_argument('--scenario', default='mixed',
                        choices=['mixed', 'viewer-random', 'gallery', 'api-only', 'stress'],
                        help='Test scenario (default: mixed)')
    parser.add_argument('--timeout', type=int, default=30,
                        help='Request timeout in seconds (default: 30)')
    parser.add_argument('--no-warmup', action='store_true',
                        help='Skip warmup phase')

    args = parser.parse_args()

    print(f"\n{'='*80}")
    print(f"MEDIA SERVER LOAD TEST")
    print(f"{'='*80}")
    print(f"URL:       {args.url}")
    print(f"Workers:   {args.workers}")
    print(f"Duration:  {args.duration}s" + (" (or until request count)" if args.requests else ""))
    print(f"Scenario:  {args.scenario}")
    print(f"Timeout:   {args.timeout}s")
    print(f"{'='*80}\n")

    # Initialize tester
    tester = LoadTester(args.url, timeout=args.timeout)

    # Warmup phase
    if not args.no_warmup:
        print("🔥 Warmup phase...")
        tester.discover_categories()
        for category in tester.categories[:3]:  # Warmup first 3 categories
            tester.fetch_file_list(category)
        print("✓ Warmup complete\n")

    # Start load test
    print(f"🚀 Starting load test with {args.workers} workers...")
    start_time = time.time()

    # Execute load test with thread pool
    with concurrent.futures.ThreadPoolExecutor(max_workers=args.workers) as executor:
        futures = [
            executor.submit(worker_thread, tester, args.scenario, args.duration, args.requests)
            for _ in range(args.workers)
        ]

        # Wait for completion
        concurrent.futures.wait(futures)

        # Collect results
        all_results = TestResults()
        for future in futures:
            worker_stats = future.result()
            for stat in worker_stats:
                all_results.add_request(stat)

    elapsed_seconds = time.time() - start_time

    # Print results
    print_results(all_results, elapsed_seconds)

    # Exit with appropriate code
    sys.exit(0 if all_results.broken_pipes == 0 else 1)


if __name__ == '__main__':
    main()

File: media-server/load_test.pl
#!/usr/bin/env perl
use strict;
use warnings;
use v5.40;

=head1 NAME

load_test.pl - Media Server Load Testing Script (Perl Implementation)

=head1 SYNOPSIS

    perl load_test.pl --url http://localhost:8080 --workers 5 --duration 60
    perl load_test.pl --scenario stress --workers 10 --duration 120
    perl load_test.pl --scenario api-only --requests 1000

=head1 DESCRIPTION

Perl implementation of load testing for media server. Validates:
- Template serialization bottleneck elimination
- Containerization readiness
- API endpoint performance
- Cross-language HTTP client behavior

=cut

use Getopt::Long qw(:config no_ignore_case bundling);
use Time::HiRes qw(time sleep);
use HTTP::Tiny;
use JSON::PP;
use URI::Encode qw(uri_encode);
use List::Util qw(min max sum shuffle);
use POSIX qw(strftime);

# Configuration
my %config = (
    url      => 'http://localhost:8080',
    workers  => 5,
    duration => 60,
    requests => undef,
    scenario => 'mixed',
    timeout  => 30,
    warmup   => 1,
);

# Results tracking
my %results = (
    total_requests      => 0,
    successful_requests => 0,
    failed_requests     => 0,
    broken_pipes        => 0,
    timeouts            => 0,
    total_duration_ms   => 0,
    min_duration_ms     => 1e9,
    max_duration_ms     => 0,
    errors_by_type      => {},
    requests_by_endpoint => {},
    durations           => [],
);

# Test data
my @categories;
my %files_by_category;

sub print_banner {
    print "\n" . "=" x 80 . "\n";
    print "MEDIA SERVER LOAD TEST (Perl $^V)\n";
    print "=" x 80 . "\n";
    printf "URL:       %s\n", $config{url};
    printf "Workers:   %d\n", $config{workers};
    printf "Duration:  %ds%s\n", $config{duration},
           (defined $config{requests} ? " (or until request count)" : "");
    printf "Scenario:  %s\n", $config{scenario};
    printf "Timeout:   %ds\n", $config{timeout};
    print "=" x 80 . "\n\n";
}

sub make_request {
    my ($http, $endpoint, $params) = @_;
    my $start = time();

    my $url = $config{url} . $endpoint;
    if ($params && %$params) {
        my @query = map { "$_=" . uri_encode($params->{$_}) } keys %$params;
        $url .= '?' . join('&', @query);
    }

    my $response = $http->get($url);
    my $duration_ms = (time() - $start) * 1000;

    return {
        endpoint    => $endpoint,
        duration_ms => $duration_ms,
        status_code => $response->{status},
        error       => $response->{success} ? undef : $response->{reason},
        content     => $response->{content},
    };
}

sub record_stat {
    my ($stat) = @_;

    $results{total_requests}++;
    $results{requests_by_endpoint}{$stat->{endpoint}}++;

    if ($stat->{error}) {
        $results{failed_requests}++;
        $results{errors_by_type}{$stat->{error}}++;

        if ($stat->{error} =~ /broken pipe/i) {
            $results{broken_pipes}++;
        }
        if ($stat->{error} =~ /timeout/i) {
            $results{timeouts}++;
        }
    } else {
        $results{successful_requests}++;
        $results{total_duration_ms} += $stat->{duration_ms};
        $results{min_duration_ms} = min($results{min_duration_ms}, $stat->{duration_ms});
        $results{max_duration_ms} = max($results{max_duration_ms}, $stat->{duration_ms});
        push @{$results{durations}}, $stat->{duration_ms};
    }
}

sub discover_categories {
    my ($http) = @_;

    # Use common test categories
    @categories = ('All', '5-★★★★★', '📁 Classico');
    print "✓ Using test categories: " . join(', ', @categories) . "\n";
    return @categories;
}

sub fetch_file_list {
    my ($http, $category) = @_;

    my $stat = make_request($http, '/api/filelist', { category => $category });

    if (!$stat->{error} && $stat->{status_code} == 200) {
        my $files = eval { decode_json($stat->{content}) };
        if ($files && ref($files) eq 'ARRAY') {
            $files_by_category{$category} = $files;
            printf "✓ Fetched %d files for category '%s'\n", scalar(@$files), $category;
            return $files;
        }
    }

    warn "⚠ Failed to fetch files for '$category'\n";
    return [];
}

sub viewer_request {
    my ($http, $category, $random_mode) = @_;

    my $files = $files_by_category{$category} || [];
    my $encoded_cat = uri_encode($category);

    if (!@$files) {
        return make_request($http, "/view/$encoded_cat");
    }

    my $file_path = $files->[int(rand(@$files))];
    my $params = { file => $file_path };
    $params->{random} = 'true' if $random_mode;

    return make_request($http, "/view/$encoded_cat", $params);
}

sub gallery_request {
    my ($http, $category, $page) = @_;
    my $encoded_cat = uri_encode($category);
    return make_request($http, "/tag/$encoded_cat", { page => $page || 1 });
}

sub api_filelist_request {
    my ($http, $category) = @_;
    return make_request($http, '/api/filelist', { category => $category });
}

sub homepage_request {
    my ($http) = @_;
    return make_request($http, '/');
}

sub mixed_workload {
    my ($http) = @_;

    return homepage_request($http) unless @categories;

    my $category = $categories[int(rand(@categories))];
    my $rand = rand(100);

    if ($rand < 40) {
        # 40% viewer requests
        return viewer_request($http, $category, int(rand(2)));
    } elsif ($rand < 70) {
        # 30% gallery requests
        return gallery_request($http, $category, int(rand(5)) + 1);
    } elsif ($rand < 90) {
        # 20% API requests
        return api_filelist_request($http, $category);
    } else {
        # 10% homepage
        return homepage_request($http);
    }
}

sub worker_thread {
    my ($scenario, $duration_seconds, $requests_count) = @_;

    my $http = HTTP::Tiny->new(
        timeout => $config{timeout},
        agent   => 'MediaServerLoadTest-Perl/1.0',
    );

    my @local_stats;
    my $start_time = time();
    my $request_count = 0;

    while (1) {
        # Check termination conditions
        my $elapsed = time() - $start_time;
        last if $duration_seconds > 0 && $elapsed >= $duration_seconds;
        last if defined($requests_count) && $request_count >= $requests_count;

        # Execute request based on scenario
        my $stat;
        if ($scenario eq 'mixed') {
            $stat = mixed_workload($http);
        } elsif ($scenario eq 'viewer-random') {
            my $cat = @categories ? $categories[int(rand(@categories))] : 'All';
            $stat = viewer_request($http, $cat, 1);
        } elsif ($scenario eq 'gallery') {
            my $cat = @categories ? $categories[int(rand(@categories))] : 'All';
            $stat = gallery_request($http, $cat);
        } elsif ($scenario eq 'api-only') {
            my $cat = @categories ? $categories[int(rand(@categories))] : 'All';
            $stat = api_filelist_request($http, $cat);
        } elsif ($scenario eq 'stress') {
            $stat = mixed_workload($http);
        } else {
            $stat = mixed_workload($http);
        }

        push @local_stats, $stat;
        $request_count++;

        # Add small delay for non-stress scenarios
        sleep(0.1) if $scenario ne 'stress' && $duration_seconds > 0;
    }

    return \@local_stats;
}

sub calculate_percentile {
    my ($sorted_array, $percentile) = @_;
    return 0 unless @$sorted_array;
    my $index = int($percentile * @$sorted_array);
    $index = @$sorted_array - 1 if $index >= @$sorted_array;
    return $sorted_array->[$index];
}

sub print_results {
    my ($elapsed_seconds) = @_;

    print "\n" . "=" x 80 . "\n";
    print "LOAD TEST RESULTS\n";
    print "=" x 80 . "\n";

    print "\n📊 Overall Statistics:\n";
    printf "  Total Requests:      %d\n", $results{total_requests};
    my $success_rate = $results{total_requests} > 0
        ? ($results{successful_requests} / $results{total_requests}) * 100
        : 0;
    printf "  Successful:          %d (%.2f%%)\n", $results{successful_requests}, $success_rate;
    printf "  Failed:              %d\n", $results{failed_requests};
    printf "  Broken Pipes:        %d %s\n", $results{broken_pipes},
           $results{broken_pipes} > 0 ? '⚠️  BOTTLENECK!' : '✅';
    printf "  Timeouts:            %d\n", $results{timeouts};
    printf "  Test Duration:       %.2fs\n", $elapsed_seconds;
    my $throughput = $elapsed_seconds > 0
        ? $results{total_requests} / $elapsed_seconds
        : 0;
    printf "  Throughput:          %.2f req/s\n", $throughput;

    if ($results{successful_requests} > 0) {
        print "\n⏱️  Response Times:\n";
        my $avg = $results{total_duration_ms} / $results{successful_requests};
        printf "  Average:             %.2f ms\n", $avg;
        printf "  Min:                 %.2f ms\n", $results{min_duration_ms};
        printf "  Max:                 %.2f ms\n", $results{max_duration_ms};

        # Calculate percentiles
        my @sorted_durations = sort { $a <=> $b } @{$results{durations}};
        if (@sorted_durations) {
            my $p50 = calculate_percentile(\@sorted_durations, 0.50);
            my $p95 = calculate_percentile(\@sorted_durations, 0.95);
            my $p99 = calculate_percentile(\@sorted_durations, 0.99);
            printf "  p50:                 %.2f ms\n", $p50;
            printf "  p95:                 %.2f ms\n", $p95;
            printf "  p99:                 %.2f ms\n", $p99;
        }
    }

    if (%{$results{requests_by_endpoint}}) {
        print "\n🎯 Requests by Endpoint:\n";
        for my $endpoint (sort { $results{requests_by_endpoint}{$b} <=>
                                 $results{requests_by_endpoint}{$a} }
                         keys %{$results{requests_by_endpoint}}) {
            printf "  %-30s %d\n", $endpoint, $results{requests_by_endpoint}{$endpoint};
        }
    }

    if (%{$results{errors_by_type}}) {
        print "\n❌ Errors by Type:\n";
        for my $error (sort { $results{errors_by_type}{$b} <=>
                             $results{errors_by_type}{$a} }
                      keys %{$results{errors_by_type}}) {
            printf "  %-30s %d\n", $error, $results{errors_by_type}{$error};
        }
    }

    print "\n" . "=" x 80 . "\n";

    # Verdict
    if ($results{broken_pipes} > 0) {
        print "⚠️  VERDICT: BROKEN PIPES DETECTED - Template serialization bottleneck still present!\n";
    } elsif ($success_rate >= 99.0 && ($results{successful_requests} == 0 ||
             $results{total_duration_ms} / $results{successful_requests} < 500)) {
        print "✅ VERDICT: EXCELLENT - System handles load gracefully!\n";
    } elsif ($success_rate >= 95.0) {
        print "✓ VERDICT: GOOD - System performs adequately under load\n";
    } else {
        print "⚠️  VERDICT: POOR - High error rate indicates issues\n";
    }
    print "=" x 80 . "\n\n";
}

# Parse command line options
GetOptions(
    'url=s'      => \$config{url},
    'workers=i'  => \$config{workers},
    'duration=i' => \$config{duration},
    'requests=i' => \$config{requests},
    'scenario=s' => \$config{scenario},
    'timeout=i'  => \$config{timeout},
    'no-warmup'  => sub { $config{warmup} = 0 },
    'help|h'     => \my $help,
) or die "Error parsing options!\n";

if ($help) {
    print <<'HELP';
Usage: perl load_test.pl [OPTIONS]

Options:
  --url URL          Base URL of media server (default: http://localhost:8080)
  --workers N        Number of concurrent workers (default: 5)
  --duration SEC     Test duration in seconds (default: 60)
  --requests N       Number of requests per worker (overrides duration)
  --scenario NAME    Test scenario: mixed, viewer-random, gallery, api-only, stress
  --timeout SEC      Request timeout in seconds (default: 30)
  --no-warmup        Skip warmup phase
  --help, -h         Show this help

Scenarios:
  mixed          Mixed workload (default) - homepage, gallery, viewer, API
  viewer-random  Focus on viewer with random mode (stress test serialization)
  gallery        Gallery page requests with pagination
  api-only       API endpoint testing (/api/filelist)
  stress         Aggressive stress test with no delays

Examples:
  perl load_test.pl --workers 5 --duration 60
  perl load_test.pl --scenario viewer-random --workers 10 --duration 30
  perl load_test.pl --scenario stress --workers 20 --duration 120
HELP
    exit 0;
}

# Main execution
print_banner();

my $http = HTTP::Tiny->new(
    timeout => $config{timeout},
    agent   => 'MediaServerLoadTest-Perl/1.0',
);

# Warmup phase
if ($config{warmup}) {
    print "🔥 Warmup phase...\n";
    discover_categories($http);
    for my $category (@categories[0..min(2, $#categories)]) {
        fetch_file_list($http, $category);
    }
    print "✓ Warmup complete\n\n";
}

# Start load test
printf "🚀 Starting load test with %d workers...\n", $config{workers};
my $start_time = time();

# Fork workers
my @pids;
my @temp_files;

for my $worker_id (1..$config{workers}) {
    my $temp_file = "/tmp/loadtest_perl_$$" . "_$worker_id.json";
    push @temp_files, $temp_file;

    my $pid = fork();
    die "Fork failed: $!" unless defined $pid;

    if ($pid == 0) {
        # Child process
        my $stats = worker_thread($config{scenario}, $config{duration}, $config{requests});

        # Write results to temp file
        open my $fh, '>', $temp_file or die "Cannot write $temp_file: $!";
        print $fh encode_json($stats);
        close $fh;

        exit 0;
    }

    push @pids, $pid;
}

# Wait for all workers
waitpid($_, 0) for @pids;

# Collect results
for my $temp_file (@temp_files) {
    next unless -f $temp_file;

    open my $fh, '<', $temp_file or next;
    my $content = do { local $/; <$fh> };
    close $fh;

    my $stats = eval { decode_json($content) };
    if ($stats && ref($stats) eq 'ARRAY') {
        record_stat($_) for @$stats;
    }

    unlink $temp_file;
}

my $elapsed_seconds = time() - $start_time;

# Print results
print_results($elapsed_seconds);

# Exit with appropriate code
exit($results{broken_pipes} > 0 ? 1 : 0);
